impl/spanner/directory/directory_test.go

```diff
@@ -0,0 +1,240 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package directory
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/keytransparency/core/directory"
+	"github.com/google/keytransparency/impl/spanner/testutil"
+	"github.com/google/trillian/crypto/keyspb"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	ktspanner "github.com/google/keytransparency/impl/spanner"
+	tpb "github.com/google/trillian"
+)
+
+// Verify that Table implements the directory.Storage interface.
+var _ directory.Storage = &Table{}
+
+func NewForTest(ctx context.Context, t *testing.T) (directory.Storage, func(context.Context)) {
+	t.Helper()
+	ddl, err := ktspanner.ReadDDL()
+	if err != nil {
+		t.Fatal(err)
+	}
+	client, cleanup := testutil.CreateDatabase(ctx, t, ddl)
+	s := New(ktspanner.NewForTest(client))
+	done := func(_ context.Context) { cleanup() }
+	return s, done
+}
+
+func TestList(t *testing.T) {
+	ctx := context.Background()
+	admin, cleanup := NewForTest(ctx, t)
+	defer cleanup(ctx)
+
+	directories := []*directory.Directory{
+		{
+			DirectoryID: "directory1",
+			Map:         &tpb.Tree{TreeId: 1},
+			Log:         &tpb.Tree{TreeId: 2},
+			VRF:         &keyspb.PublicKey{Der: []byte("pubkeybytes")},
+			VRFPriv:     &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+			MinInterval: 1 * time.Second,
+			MaxInterval: 5 * time.Second,
+		},
+		{
+			DirectoryID: "directory2",
+			Map:         &tpb.Tree{TreeId: 1},
+			Log:         &tpb.Tree{TreeId: 2},
+			VRF:         &keyspb.PublicKey{Der: []byte("pubkeybytes")},
+			VRFPriv:     &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+			MinInterval: 5 * time.Hour,
+			MaxInterval: 500 * time.Hour,
+			Deleted:     true,
+		},
+	}
+	for _, d := range directories {
+		if err := admin.Write(ctx, d); err != nil {
+			t.Errorf("Write(): %v", err)
+			continue
+		}
+		if err := admin.SetDelete(ctx, d.DirectoryID, d.Deleted); err != nil {
+			t.Errorf("SetDelete(%v, %v): %v", d.DirectoryID, d.Deleted, err)
+			continue
+		}
+	}
+
+	for _, tc := range []struct {
+		readDeleted bool
+		want        []*directory.Directory
+	}{
+		{
+			readDeleted: true,
+			want:        directories,
+		},
+		{
+			readDeleted: false,
+			want:        []*directory.Directory{directories[0]},
+		},
+	} {
+		directories, err := admin.List(ctx, tc.readDeleted)
+		if err != nil {
+			t.Errorf("List(): %v", err)
+			continue
+		}
+
+		if got, want := len(directories), len(tc.want); got != want {
+			t.Errorf("Got %v directories, want %v", got, want)
```

Comment:
```suggestion
			t.Fatalf("Got %v directories, want %v", got, want)
```
Otherwise the test can panic with out-of-bound in the loop below.